### ISSUE TITLE
LoadBalancer forwardingRules order

### ISSUE THREAD

**Issue by rienafairefr** [pulumi/pulumi-digitalocean#85]
pulumi version: 1.14.1
pulumi-digitalocean version: 1.6.0
SDK: python

`pulumi up`-ing a stack that contains a DO load balancer, I was surprised it needed to update the load balancer, because I didn't modify anything for that resource. The details show that it considers the order of the rules:

    ~ digitalocean:index/loadBalancer:LoadBalancer: (update)
        [id=4223f162-a5c0-4ad0-82c2-e3c54befe702]
        [urn=urn:pulumi:xxx::yyy::digitalocean:index/loadBalancer:LoadBalancer::loadbalancer]
        [provider=urn:pulumi:xxx::yyy::pulumi:providers:digitalocean::default_1_6_0::949ace48-ce62-47da-9f42-59dcbd161414]
      ~ forwardingRules: [
          ~ [0]: {
                  + certificateId: "49bb2cd0-96e7-4916-a7bd-9b61f35f47fa"
                  ~ entryPort    : 80 => 443
                  ~ entryProtocol: "http" => "https"
                }
          ~ [1]: {
                  - certificateId: "49bb2cd0-96e7-4916-a7bd-9b61f35f47fa"
                  ~ entryPort    : 443 => 80
                  ~ entryProtocol: "https" => "http"
                }
        ]

It basically wants to exchange element #0 and element #1 in a list, the order of it shouldn't matter, this is weird.

**Comment by rienafairefr** [comment]
Note: the order seems to ping-pong, it's updating the loadbalancer on each update, exhanging element 0 and 1 each time :-)

**Comment by leezen** [comment]
Possibly related: https://github.com/pulumi/pulumi-aws/issues/920

**Comment by lukehoban** [comment]
@rienafairefr Any chance you can share a repro for this?

The upstream provider implements [list semantics](https://github.com/terraform-providers/terraform-provider-digitalocean/blob/master/digitalocean/resource_digitalocean_loadbalancer.go#L58), not set semantics, for the `forwarding_rules` on this resource.  And it does not internally do any ordering - it just uses what comes back from DO. 
 Which means that order does matter in the user program.

I tried this myself, and as long as the order I have the rules specified matches the oder DO's REST API reports them, I do not see a diff in Pulumi.  If the order I specify them is different, Pulumi and the upstream provider accept trying to make the change, but no change is actually made in DO itself.  

From what I can tell - these are not logically ordered in DO (and requesting them in one order, the DO API will return back in a different order).  If so, then the upstream Terraform provider likely needs to use set semantics.

**Comment by rienafairefr** [comment]
A repro, we can probably shave off some more, but using this:
```
import pulumi
import pulumi_digitalocean as do

domain = pulumi.config.get_config('repro:domain')

certificate = do.Certificate(
    f'certificate',
    type='lets_encrypt',
    domains=[f'invalid.{domain}']
)

load_balancer = do.LoadBalancer(
    'loadbalancer',
    redirect_http_to_https=True,
    region='ams3',
    forwarding_rules=[
        {
            'certificate_id': certificate.id,
            'entryPort': 443,
            'targetProtocol': 'http',
            'entryProtocol': 'https',
            'targetPort': 80
        },
        {
            'entryPort': 80,
            'targetProtocol': 'http',
            'entryProtocol': 'http',
            'targetPort': 80
        }
    ]
)
```
`pulumi up` will want to do an update every time, I thought it would "ping-pong" as I said above, but in fact no, it just can't change the order of the rules, even if it wants to.

OK, I understand the list semantics and I think you've pinpointed the problem. So.... in order to not have this problem, the list that is built in the pulumi program should be in the same order as the order that digitalocean would report back. This can be easily fixed for short lists, but I think people might have more complex forwarding rules than just two items in a list.

I would personally classify that as a bug in upstream (terraform, or digitalocean API), either the forwarding_rule terraform field should have a set semantics  (order shouldn't matter) ? Or the DO API should honor the requests to change the order of the forwarding rules. 

As a quick fix for the repro above
```
    forwarding_rules=[
        {
            'entryPort': 80,
            'targetProtocol': 'http',
            'entryProtocol': 'http',
            'targetPort': 80
        },
        {
            'certificate_id': certificate.id,
            'entryPort': 443,
            'targetProtocol': 'http',
            'entryProtocol': 'https',
            'targetPort': 80
        },
    ]
```
doesn't need updating every time :+1:

**Comment by rienafairefr** [comment]
I can confirm that the issue is reproducible using pure terraform, as expected :-)

**Comment by lukehoban** [comment]
Great - thanks @rienafairefr.  I'll close this out as a dupe of  https://github.com/terraform-providers/terraform-provider-digitalocean/issues/414 then.

**Comment by rienafairefr** [comment]
The terraform [upsteam issue](https://github.com/terraform-providers/terraform-provider-digitalocean/issues/414) has been solved, what's the process so that it gets down to pulumi, the fix is released in terraform digitalocean provider 1.17.0 ?
